src/GeositeFramework/js/Map.js

```diff
@@ -201,132 +201,88 @@ require(['use!Geosite',
-            // visible legend elements.
-            view.$legendEl.find('.esriLegendService').each(function(idx, legendParent) {
-                var $groupLayers = $(legendParent).find('.esriLegendGroupLayer:visible');
-                if ($groupLayers.length > 0) {
-                    legendNuggets.push($groupLayers);
-                    nuggetCount += $groupLayers.length;
-                } else {
-                    // Tile layer legends are rendered outside of a .esriLegendService element.
-                    legendNuggets.push($(legendParent).find('div:visible'));
-                    nuggetCount += 1;
+                if (serviceInfo && service.visible && serviceInfo.pluginObject.showServiceLayersInLegend) {
+                    service.visibleLayers.sort();
+                    _.each(service.visibleLayers, function(layerId) {
+                        var layer = _.findWhere(service.layerInfos, {id: layerId});
+                        if (!layer) {
+                            return;
+                        }
+                        var serviceLegend = getServiceLegend(service.url);
+                        if (!serviceLegend) {
+                            return;
+                        }
+                        var legend = _.findWhere(serviceLegend, {layerId: layerId});
+                        if (isLayerInScale(service, layer)) {
+                            result.push({
+                                service: service,
+                                layer: layer,
+                                legend: legend
+                            });
+                        }
+                    });
                 }
             });
+            return result;
+        }
 
-            legendNuggets.sort(heightComparator);
-
-            view.$legendEl.empty()
-                .append.apply(view.$legendEl, legendNuggets);
-
-            // Compare the total number of visible layers on the map to the number
-            // of legendNuggets to determine if the legend has gotten info for all
-            // visible layers. If not, recur with a delay.
-            //
-            // This is a hack, because sometimes when loading from browser cache,
-            // layerAdd events don't fire.
-
-            // Get all the layers on the map, and filter out graphic layers (i.e. subregion boundaries)
-            // and layers that aren't visible;
-            var totalVisibleLayers =_.filter(esriMap.getLayersVisibleAtScale(), function(layer) {
-                return !layer.graphics && layer.visible;
-            });
-
-            // Subtract one for the base layer.
-            if ((totalVisibleLayers.length - 1) !== nuggetCount) {
-                _.delay(updateLegend, 250);
-            }
+        // Filter out layers that are not visible at the current map scale.
+        // Adapted from the ESRI dijit legend source code. (Ref: _isLayerInScale)
+        function isLayerInScale(service, layer) {
```

Comment:
This may not be helpful, but there's a method on `map` to return visible layers: [getLayersVisibleAtScale](https://developers.arcgis.com/javascript/jsapi/map.html#getlayersvisibleatscale).